pkg/controller/elasticsearch/reconcile/status.go

```diff
@@ -191,20 +194,23 @@ func (u *UpgradeReporter) Merge(other esv1.UpgradeOperation) esv1.UpgradeOperati
 	if u == nil {
 		return *upgradeOperation
 	}
-	nodes := make([]esv1.UpgradedNode, 0, len(u.nodes))
-	for _, node := range u.nodes {
-		nodes = append(nodes, esv1.UpgradedNode{
-			Name:      node.Name,
-			Predicate: node.Predicate,
-			Message:   node.Message,
-			Status:    node.Status,
+	var nodes []esv1.UpgradedNode
+	if len(u.nodes) != 0 {
+		nodes = make([]esv1.UpgradedNode, 0, len(u.nodes))
+		for _, node := range u.nodes {
+			nodes = append(nodes, esv1.UpgradedNode{
+				Name:      node.Name,
+				Predicate: node.Predicate,
+				Message:   node.Message,
+				Status:    node.Status,
+			})
+		}
+		// Sort for stable comparison
+		sort.Slice(nodes, func(i, j int) bool {
+			return nodes[i].Name < nodes[j].Name
 		})
 	}
-	// Sort for stable comparison
-	sort.Slice(nodes, func(i, j int) bool {
-		return nodes[i].Name < nodes[j].Name
-	})
-	if (u.nodes != nil && !reflect.DeepEqual(nodes, other.Nodes)) || upgradeOperation.Nodes == nil {
+	if (u.nodes != nil && !reflect.DeepEqual(nodes, other.Nodes)) || upgradeOperation.LastUpdatedTime.IsZero() {
```

Comment:
I would add the following kind of unit test to ensure that the status is updated only when the reporter has been used (`u.nodes != nil`):

```golang
		{
			name: "Do not overwrite existing status if reporter is not used",
			state: func() *State {
				s := MustNewState(esv1.Elasticsearch{})
				return s
			},
			args: args{
				otherStatus: esv1.ElasticsearchStatus{
					Phase: esv1.ElasticsearchResourceInvalid,
					Conditions: esv1.Conditions{
						{
							Type:    "ReconciliationComplete",
							Status:  "False",
							Message: "Nodes upgrade in progress",
						},
					},
					InProgressOperations: esv1.InProgressOperations{
						DownscaleOperation: esv1.DownscaleOperation{
							Nodes: nil,
						},
						UpgradeOperation: esv1.UpgradeOperation{
							LastUpdatedTime: metav1.Now(),
							Nodes: []esv1.UpgradedNode{
								{
									Message:   pointer.String("Cannot restart node because of failed predicate"),
									Name:      "node-1",
									Status:    "PENDING",
									Predicate: pointer.String("if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas"),
								},
							},
						},
						UpscaleOperation: esv1.UpscaleOperation{
							Nodes: nil,
						},
					},
				},
			},
			wantElasticsearchStatus: esv1.ElasticsearchStatus{
				Phase: esv1.ElasticsearchResourceInvalid,
				Conditions: esv1.Conditions{
					{
						Type:    "ReconciliationComplete",
						Status:  "False",
						Message: "Nodes upgrade in progress",
					},
				},
				InProgressOperations: esv1.InProgressOperations{
					DownscaleOperation: esv1.DownscaleOperation{
						Nodes: nil,
					},
					UpgradeOperation: esv1.UpgradeOperation{
						Nodes: []esv1.UpgradedNode{
							{
								Message:   pointer.String("Cannot restart node because of failed predicate"),
								Name:      "node-1",
								Status:    "PENDING",
								Predicate: pointer.String("if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas"),
							},
						},
					},
					UpscaleOperation: esv1.UpscaleOperation{
						Nodes: nil,
					},
				},
			},
			wantPendingNewNodes: false,
		}
```